src/stdlib_stats_distribution_PRNG.fypp

```diff
@@ -0,0 +1,151 @@
+
+
+    contains
+
+    #:for k1, t1 in INT_KINDS_TYPES
+    function dist_rand_${t1[0]}$${k1}$(n) result(res)
+    !! Random integer generation for various kinds
+    !! result = [-2^k, 2^k - 1], k = 7, 15, 31, 63, depending on input kind
+    !! Result will be operated by bitwise operators to generate desired integer
+    !! and real pseudorandom numbers
+    !!
+        ${t1}$, intent(in) :: n
+        ${t1}$ :: res
+        integer :: k
+
+        k = MAX_INT_BIT_SIZE - bit_size(n)
+        if(k < 0) call error_stop("Error(dist_rand): Integer bit size is"       &
+                  //" greater than 64bit")
+        res = shiftr(xoshiro256ss( ), k)
+    end function dist_rand_${t1[0]}$${k1}$
+
+    #:endfor
+
+    function xoshiro256ss( ) result (res)
+    ! Generate random 64-bit integers
+    ! Written in 2018 by David Blackman and Sebastiano Vigna (vigna@acm.org)
+    ! http://prng.di.unimi.it/xoshiro256starstar.c
+    !
+    ! This is xoshiro256** 1.0, one of our all-purpose, rock-solid
+    ! generators. It has excellent (sub-ns) speed, a state (256 bits) that is
+    ! large enough for any parallel application, and it passes all tests we
+    ! are aware of.
+    !
+    ! The state must be seeded so that it is not everywhere zero. If you have
+    ! a 64-bit seed, we suggest to seed a splitmix64 generator and use its
+    ! output to fill st.
+    !
+    ! Fortran 90 version translated from C by Jim-215-Fisher
+    !
+        integer(int64) :: res, t
+
+        if(.not. seed_initialized) call random_distribution_seed_iint64(si,t)
+        res = rol64(st(2) * 5 , 7) * 9
+        t = shiftl(st(2), 17)
+        st(3) = ieor(st(3), st(1))
+        st(4) = ieor(st(4), st(2))
+        st(2) = ieor(st(2), st(3))
+        st(1) = ieor(st(1), st(4))
+        st(3) = ieor(st(3), t)
+        st(4) = rol64(st(4), 45)
+    end function xoshiro256ss
+
+    function rol64(x, k) result(res)
+        integer(int64), intent(in) :: x
+        integer, intent(in) :: k
+        integer(int64) :: t1, t2, res
+
+        t1 = shiftr(x, (64 - k))
+        t2 = shiftl(x, k)
+        res = ior(t1, t2)
+    end function rol64
+
+
+    function splitmix64(s) result(res)
+    ! Written in 2015 by Sebastiano Vigna (vigna@acm.org)
+    ! This is a fixed-increment version of Java 8's SplittableRandom
+    ! generator.
+    ! See http://dx.doi.org/10.1145/2714064.2660195 and
+    ! http://docs.oracle.com/javase/8/docs/api/java/util/SplittableRandom.html
+    !
+    ! It is a very fast generator passing BigCrush, and it can be useful if
+    ! for some reason you absolutely want 64 bits of state.
+    !
+    ! Fortran 90 translated from C by Jim-215-Fisher
+    !
+        integer(int64) :: res, int01, int02, int03
+        integer(int64), intent(in), optional :: s
+        data int01, int02, int03/-7046029254386353131_int64, &
+                                 -4658895280553007687_int64, &
+                                 -7723592293110705685_int64/
```

Comment:
These could be rewritten as `integer(int64), parameter`, which IMO would be cleaner, and I think just as efficient as `data` statement.